### PR TITLE
test(pallets): add `instantiate_module_*` benches per section

### LINKED ISSUE #3907: Investigate whether we need to charge for data section
Presently we have a benchmark, which calculates weight for module instantiation depending on code size:
https://github.com/gear-tech/gear/blob/e1fb7082d27a7296dd84c1b916c1ab2c65945b36/pallets/gear/src/benchmarking/mod.rs#L375-L383

Here we charge for instantiation (not the best naming yes):
https://github.com/gear-tech/gear/blob/e1fb7082d27a7296dd84c1b916c1ab2c65945b36/pallets/gear/src/queue.rs#L159-L163

The problem is that code size is summary size of all WASM code sections. But in benchmark we increase only _code section_, so actually it's time for wasm functions code instantiation. On other hand we can increase _data section_ and check what will happen. Maybe _data section_ instantiation is heavier than code instantiation, so here we have potential problem with `gear::run` time execution. Maybe _data section_ instantiation weight is significantly smaller than _code section_ instantiation weight, so here we have potential weight gain in case we will charge for _code section_ size and _data section_ size separately.

What can you do:
1) you can run instantiation benchmark and see the results:
```bash
$ cargo build -p gear-cli --release --features=runtime-benchmarks
$ ./target/release/gear benchmark pallet --chain=vara-dev --pallet=pallet_gear  --extrinsic=instantiate_module_per_kb
```
2) change `instantiate_module_per_kb`, so that two params will be taken into account, see two params example in `gr_create_program_per_kb`.
3) analyse the results and think what can be done here

### LINKED ISSUE #3942: Investigate how other types of sections (excluding code and data) in a wasm module contribute to instantiation time
### Problem to Solve

Investigate how other types of sections (excluding code and data) in a wasm module contribute to instantiation time.
Section like:
 * type
 * table
 * global 
 * etc..

### Possible Solution

Investigate by writing instantiation module benches for these sections.

### PR DESCRIPTION
**Pull request by ByteNacked**
Closes #3907 .
Closes #3942 .

Add `instantiate_module_*` benchmarks, namely:
* instantiate_module_data_section_per_kb
* instantiate_module_global_section_per_kb
* instantiate_module_table_section_per_kb
* instantiate_module_type_section_per_kb

@reviewer-or-team

### CODE REVIEW

File: pallets/gear/src/benchmarking/mod.rs [hunk 1]
```diff
@@ -113,6 +113,7 @@ const MAX_PAYLOAD_LEN: u32 = 32 * 64 * 1024;
 const MAX_PAYLOAD_LEN_KB: u32 = MAX_PAYLOAD_LEN / 1024;
 const MAX_PAGES: u32 = 512;
 const MAX_SALT_SIZE_BYTES: u32 = 4 * 1024 * 1024;
+const MAX_NUMBER_OF_DATA_SEGMENTS: u32 = 1024;
```
Comment (ByteNacked) [inline pallets/gear/src/benchmarking/mod.rs:116]: See https://github.com/gear-tech/gear/issues/3907#issuecomment-2083556107

**Review by grishasobol — COMMENTED** [review]
Please update weights files adding corresponding weight in this PR.

**Comment by ByteNacked** [comment]
> Please update weights files adding corresponding weight in this PR.

Fixed

**Review by ukint-vs — APPROVED** [review]

**Comment by ByteNacked** [comment]
> Overall looks good, but how would they be used and when?

It will be used in this draft #3937 when its ready

**Comment by ByteNacked** [comment]
* Add additional bench per section instantiation\
* Updated weights

File: pallets/gear/src/benchmarking/mod.rs [hunk 2]
```diff
@@ -384,6 +385,46 @@ benchmarks! {
         Environment::new(ext, &code, DispatchKind::Init, Default::default(), max_pages::<T>().into()).unwrap();
     }
 
+    // `d`: Size of the data section in kilobytes.
+    instantiate_module_data_section_per_kb {
+        let d in 0 .. T::Schedule::get().limits.code_len / 1024;
+
+        let WasmModule { code, .. } = WasmModule::<T>::sized_data_section(d * 1024, MAX_NUMBER_OF_DATA_SEGMENTS);
+    }: {
+        let ext = Externalities::new(ProcessorContext::new_mock());
```
Comment (breathx) [inline pallets/gear/src/benchmarking/mod.rs:394]: ext creation should be in upper section for each case

Comment (breathx) [reply pallets/gear/src/benchmarking/mod.rs:394]: just to split it logically "preparations" | "benchmarking" | "verification"

File: pallets/gear/src/benchmarking/code.rs [hunk 3]
```diff
@@ -315,7 +328,24 @@ where
...
 
-        let code = program.build();
+        let mut code = program.build();
+
+        // Add dummy type section
+        if let Some(types) = def.types {
+            for section in code.sections_mut() {
+                if let Section::Type(sec) = section {
+                    for _ in 0..types.num_elements {
+                        sec.types_mut().push(Type::Function(FunctionType::new(
```
Comment (breathx) [inline pallets/gear/src/benchmarking/code.rs:338]: pretty interesting that its still valid wasm module

**Review by breathx — APPROVED** [review]

File: pallets/gear/src/benchmarking/code.rs (hunk 3 above)
Comment (ByteNacked) [reply pallets/gear/src/benchmarking/code.rs:338]: spec doesn't seem to forbid duplicated and unused types, so it seems logical.

Comment (breathx) [reply pallets/gear/src/benchmarking/code.rs:338]: yeah, sure, I was just sharing some thoughts)

**Comment by breathx** [comment]
Ready to be merged @grishasobol ?

File: pallets/gear/src/benchmarking/code.rs [hunk 4]
```diff
@@ -378,6 +408,92 @@ where
...
+        // Maximum size of encoded i64 global is 14 bytes.
+        module.num_globals = target_bytes / 14;
+        module.full_length_globals = true;
+
+        module.into()
+    }
+
+    /// Creates a WebAssembly module with a table size of `target_bytes` bytes.
+    /// Each element in the table points to function index `0` and occupies 1 byte.
+    pub fn sized_table_section(target_bytes: u32) -> Self {
```
Comment (grishasobol) [inline pallets/gear/src/benchmarking/code.rs:463]: We have only one table segment in bench? Did you investigate how amount of tables can influence on instantiation time?

**Review by grishasobol — APPROVED** [review]

File: pallets/gear/src/benchmarking/code.rs (hunk 4 above)
Comment (ByteNacked) [reply pallets/gear/src/benchmarking/code.rs:463]: Good point, will investigate the problem and provide solution in next PR

**Pull request merged**